### PR TITLE
Bug 1899565: support raw block devices

### PR DESCRIPTION
**Pull request by bennyz**
Add support for raw block PVCs

### CODE REVIEW

**Review by rgolangh — COMMENTED** [review]
overall looks good, small question inside

File: pkg/service/node.go [hunk 1]
```diff
@@ -79,6 +86,10 @@ func (n *NodeService) NodeUnstageVolume(ctx context.Context, req *csi.NodeUnstag
 }
 
 func (n *NodeService) NodePublishVolume(ctx context.Context, req *csi.NodePublishVolumeRequest) (*csi.NodePublishVolumeResponse, error) {
+	if req.VolumeCapability == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "Missing volume capability")
```
[the remark this answers is not in the document]
Comment (bennyz) [reply pkg/service/node.go:90]: probably never, I'll remove

File: pkg/service/node.go [hunk 2]
```diff
@@ -91,6 +102,10 @@ func (n *NodeService) NodePublishVolume(ctx context.Context, req *csi.NodePublis
 		return nil, err
 	}
 
+	if req.VolumeCapability.GetBlock() != nil {
+		return n.publishBlockVolume(req, device)
+	}
+
 	targetPath := req.GetTargetPath()
 	err = os.MkdirAll(targetPath, 0750)
```
[the remark this answers is not in the document]
Comment (bennyz) [reply pkg/service/node.go:110]: ack

File: pkg/service/node.go [hunk 3]
```diff
@@ -122,6 +133,30 @@ func (n *NodeService) NodeUnpublishVolume(ctx context.Context, req *csi.NodeUnpu
...
+		if !os.IsExist(err) {
+			return nil, status.Errorf(codes.Internal, "Failed to create targetPath %s, err: %v", req.TargetPath, err)
+		}
+	}
+
+	mounter := mount.New("")
+	err = mounter.Mount(device, req.TargetPath, "", []string{"bind"})
+	if err != nil {
+		if removeErr := os.Remove(req.TargetPath); removeErr != nil {
+			return nil, status.Errorf(codes.Internal, "Failed to remove mount target %v, err: %v", req.TargetPath, err)
```
Comment (rgolangh) [inline pkg/service/node.go:150]: here the removeErr is lost. Maybe wrap both of them?

Comment (bennyz) [reply pkg/service/node.go]: ack

File: pkg/service/node.go [hunk 4]
```diff
@@ -122,6 +133,29 @@ func (n *NodeService) NodeUnpublishVolume(ctx context.Context, req *csi.NodeUnpu
...
+		if !os.IsExist(err) {
+			return nil, status.Errorf(codes.Internal, "Failed to create targetPath %s, err: %v", req.TargetPath, err)
+		}
+	}
+
+	mounter := mount.New("")
+	err = mounter.Mount(device, req.TargetPath, "", []string{"bind"})
+	if err != nil {
+		if removeErr := os.Remove(req.TargetPath); removeErr != nil {
+			return nil, status.Errorf(codes.Internal, "Failed to remove mount target %v, err: %v", req.TargetPath, removeErr)
```
Comment (rgolangh) [inline pkg/service/node.go:150]: now the failing mount error is lost

Comment (bennyz) [reply pkg/service/node.go:150]: right, sorry, I didn't get that at first

File: pkg/service/node.go [hunk 5]
```diff
@@ -122,6 +133,29 @@ func (n *NodeService) NodeUnpublishVolume(ctx context.Context, req *csi.NodeUnpu
...
+		if !os.IsExist(err) {
+			return nil, status.Errorf(codes.Internal, "Failed to create targetPath %s, err: %v", req.TargetPath, err)
+		}
+	}
+
+	mounter := mount.New("")
+	err = mounter.Mount(device, req.TargetPath, "", []string{"bind"})
+	if err != nil {
+		if removeErr := os.Remove(req.TargetPath); removeErr != nil {
+			return nil, status.Errorf(codes.Internal, "Failed to remove mount target %v, err: %v, mount error: %v", req.TargetPath, removeErr,)
```
Comment (bennyz) [inline pkg/service/node.go:150]: is there a nicer way?

Comment (rgolangh) [reply pkg/service/node.go:150]: your missing the `err` for the ErrorF, that's it I think:
```return nil, status.Errorf(
     codes.Internal,
     "Failed to remove mount target %v, err: %v, mount error: %v",
     req.TargetPath,
     removeErr,
     err)
```

Comment (rgolangh) [reply pkg/service/node.go:150]:
```suggestion
			return nil, status.Errorf(codes.Internal, "Failed to remove mount target %v, err: %v, mount error: %v", req.TargetPath, removeErr, err)
```

Comment (bennyz) [reply pkg/service/node.go:150]: :facepalm:

Comment (bennyz) [reply pkg/service/node.go]: apparently trailing commas are allowed

**Pull request merged**